ivi-input-modules/ivi-input-controller/src/ivi-input-controller.c

```diff
@@ -470,6 +470,17 @@ keyboard_grab_modifiers(struct weston_keyboard_grab *grab, uint32_t serial,
 static void
 keyboard_grab_cancel(struct weston_keyboard_grab *grab)
 {
+    struct seat_ctx *ctx_seat = wl_container_of(grab, ctx_seat, keyboard_grab);
+    struct ivisurface *surf;
+    struct weston_surface *w_surf;
+    const struct ivi_layout_interface *interface =
+                                ctx_seat->input_ctx->ivishell->interface;
+
+    wl_list_for_each(surf, &ctx_seat->input_ctx->ivishell->list_surface,
+                     link) {
+        w_surf = interface->surface_get_weston_surface(surf->layout_surface);
+        input_ctrl_kbd_leave_surf(ctx_seat, surf, w_surf);
```

Comment:
why do you think that we have to send leave events to clients, when compositor is shutting down ? Even in default grab they are not doing it

Reply:
Ideally, we should not use any foreign resources during deinitilization. At this point, we dont know if interface is still valid.

Reply:
>why do you think that we have to send leave events to clients, when compositor is shutting down ? Even in default grab they are not doing it

for the drm backend in the seat destroy flow I see the keyboard handling doing it. In the flow drm_destroy->udev_input_destroy->udev_seat_destroy->notify_keyboard_focus_out, weston keyboard focus is set to NULL and grab is cancelled. In this patch grab is combined with something equivalent to setting the focus to NULL.

>Ideally, we should not use any foreign resources during deinitilization. At this point, we dont know if interface is still valid.

Interfaces are valid because Backend will be destroyed after the plugins and as pointed in the flow above events are still sent by defaut input grab interface of weston OR input.c.

Reply:
notify_keyboard_focus_out calls weston_keyboard_cancel_grab. But in the end it is empty function in default grab interface. See: default_grab_keyboard_cancel. You don't need to send leave events to clients. Also ivi_layout_interface could be invalid at this point.

Reply:
From the functionality point of view in notify_keyboard_focus_out call, weston_keyboard_focus is being set to NULL. Its fine if default grab is not doing, but intention is to set focus to NULL.

Reply:
you are talking about weston internal focus. Weston does not send any wl_keyboard_send_leave events to clients during shutdown.

Reply:
It does send in function weston_keyboard_set_focus. This function is called during shutdown and if there is a keyboard focus then leave will be sent (input.c:1549). Hence same behavior is followed in these patches i.e. in the absence of input_controller what weston does, similar actions are performed by input_controller.